File: 2021/09/23/CountBinarySubstrings.java
/**
 * https://leetcode.com/problems/count-binary-substrings/
 * @author Andrew Jarombek
 * @since 9/23/2021
 */

public class CountBinarySubstrings {
    public int countBinarySubstrings(String s) {
        int count = 0;
        int previous = 0;
        int current = 0;
        char currentChar = s.charAt(0);

        for (char c : s.toCharArray()) {
            if (c == currentChar) {
                current++;
            } else {
                previous = current;
                current = 1;
                currentChar = c;
            }

            if (current > 0 && previous > 0 && current <= previous) {
                count++;
            }
        }

        return count;
    }
}

File: 2022/04/10/countBinarySubstrings.ts
/**
 * https://leetcode.com/problems/count-binary-substrings/
 * @author Andrew Jarombek
 * @since 4/10/2022
 */

function countBinarySubstrings(s: string): number {
    let result = 0;
    let prev = 0;
    let current = 0;
    let value = "0";

    for (let c of s) {
        if (c != value) {
            result += Math.min(prev, current);

            value = c;
            prev = current;
            current = 1;
        } else {
            current++;
        }
    }

    return result + Math.min(prev, current);
}